Fix for wrong version processing

brejoc opened the pull request:
### What does this PR do?

Since version and epoch seem to be split by default now, the yum pkg info
processing should have changed.

* Fixes the version info processing for pkg installation.
* Test for installation with and without diff_attr.

### What issues does this PR fix or reference?
-
### Previous Behavior

Installation of packages failed due to wrong version/epoch processing.

### New Behavior

Installation of pkgs works via yum.

### Tests written?

Yes

### Commits signed with GPG?

Yes

brejoc committed:
Fix for wrong version processing

Since version and epoch seem to be split by default now, the yum pkg info
processing should have changed.

* Fixes the version info processing for pkg installation.
* Test for installation with and without diff_attr.

kkaempf commented:
Hmm, are you removing epoch handling here ? While epoch is forbidden in SUSE, Redhat/Fedora make use of it.

brejoc commented:
@kkaempf Epoch information now seems to be in a separate field. So this shouldn't break epoch handling. But perhaps it would make sense, if @moio would take a look here, since the underlying code is from him.

isbm commented on salt/modules/yumpkg.py:
```diff
@@ -1513,18 +1513,20 @@ def install(name=None,
+                    else:
+                        # No diff_attr was provided, so version is directly
+                        # in ver.
+                        ver2 = ver
```
I would write it easier and safer:
```python
ver2 = diff_attr and ver.get('version') or ver
```

moio replied:
I for one find the original version clearer - especially because your version, not being parenthesised, forces the reader to think about operator precedence.

I do agree, though, that using an expression is safer. How about:

```python
ver2 = ver['version'] if diff_attr else ver
```

dincamihai replied:
+1 for @moio

isbm replied:
@moio so it is:
```python
ver2 = ver['version'] if diff_attr else ver
```
I am fine with this.

brejoc replied:
As soon as there is a `diff_attr`, the version is added to the dict. I saw that in the code and just verified it.

isbm replied:
@brejoc OK, in `pkg_resource::format_pkg_list`, if `attr`, then `version` is pushed, yes. In this case type of the `ver` variable changes between tuple and dict. Then I would do it differently:
```python
ver2 = isinstance(ver, dict) and ver.get('version') or ver
```

brejoc replied:
Because I consider overwriting the variable used in the loop as a bad practice. It's a constant stream of failure. The loop is short, but it's a habit of mine that worked well for me.

isbm replied:
The variable is only used once. I see no need to allocate memory twice. Really.

isbm replied:
@brejoc I actually see it as a bugmaking: you supposed to extract the version, which is a tuple of things. It is either put in a structure or made directly. I agree that this design itself is very strange, but we have what we have. Now you created "var2" like this is another version. I don't like that and would prefer to get rid of what one doesn't need.

isbm commented on salt/modules/yumpkg.py:
```diff
@@ -1535,10 +1537,17 @@ def install(name=None,
+                        else:
+                            # No diff_attr was provided, so version is directly
+                            # in ver.
+                            ver2 = ver
```
Same here as above: no need to grind 9 codelines. Comments may stay.

moio commented on salt/modules/yumpkg.py:
```diff
@@ -1513,18 +1513,20 @@ def install(name=None,
             cver = old_as_list.get(pkgname, [])
             if reinstall and cver:
                 for ver in cver:
-                    ver = norm_epoch(ver, version_num)
```
Do I understand correctly that this line did not work before this PR in case case `ver` was a dict (`diff_attr == True`)?

brejoc replied:
I just stepped through the code again to double check that. We still need this part, since the version number still includes the epoch if no `diff_attr` is provided. So I was wrong on that. What failed was the splitting of `ver` in `norm_epoch`, since `ver` can be a string and a dict.

brejoc committed:
Add epoch parsing again

We still need the epoch parsing for the case where no diff_attr is provided.
In that case the epoch information is included in the version_num.

rallytime commented:
@brejoc There's some lint on this: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/17357/violations/file/tests/unit/modules/test_yumpkg.py/ :)

brejoc commented on salt/modules/yumpkg.py:
```diff
@@ -1513,18 +1513,20 @@ def install(name=None,
+                    else:
+                        # No diff_attr was provided, so version is directly
+                        # in ver.
+                        ver2 = ver
```
@isbm If I remember correctly, memory is not allocated twice, since both vars point to the same place in memory. 😜
Think of it it as an investment in future bug prevention. This code will probably only grow over time. 😄

brejoc committed:
Make linter happy

isbm commented on salt/modules/yumpkg.py:
```diff
@@ -1535,10 +1539,13 @@ def install(name=None,
+                        # if diff_attr was provided, version number is directly
+                        # in ver, otherwise its a dict.
+                        ver2 = ver['version'] if diff_attr else ver
+                        ver2 = norm_epoch(ver2, version_num)
```
Pascal programmers? :wink: 
```python
ver = norm_epoch(ver['version'] if diff_attr else ver, version_num)
```

isbm replied:
I am fine if you move all that to another function that would "extract version" or "normalise version" etc, as long this is needed again somewhere.

isbm replied:
@brejoc as of "preventing from bugs" it doesn't, but only adds to. Because I see it as a broken logic right now: in a loop you have a variable that can be A-type and B-type, while the only consumer you have can take B-type. Typically, a Java programmer will right away create another variable, because they cannot swap a type. In Python we do not need to pollute the scope with such unnecessary stuff, since we can extract data and reset the pointer where we need to, then pass to the consumer.

As of "code will grow" — when it will grow, _then_ they will add. No need to make complexity where it is not needed.

moio replied:
@isbm I think this is getting a little bit too far - it's been 10 comments now just to argue about a one-liner versus the current 2 lines that add one extra (immutable) variable.

I suggest to call this a style difference, accept the fact different people have different styles, and just end the discussion here.

:peace_symbol:

isbm replied:
@moio there is no war, so :peace_symbol: is misleading here. Merging this and showing this as a bad example to others who will copy such things "because others did this already" — is a bad idea.

I am down-voting this multiline Java coding, but if @terminalmage says "yeah OK", then let it be so.

brejoc replied:
@isbm There is something we can agree on. I also don't like the fact that `ver` can either be a string or a dict. So how about a little bit of refactoring to make sure that `cver` is always a list of dicts, even if `version` is the only key?

isbm replied:
@brejoc correction: _tuple_ or a dict. :wink: I absolutely for that refactoring then. Agreed. :stuck_out_tongue_winking_eye:

brejoc replied:
Here is the [branch](https://github.com/brejoc/salt/commits/develop-yum-install-refactoring) with the refactoring.

dincamihai commented on salt/modules/yumpkg.py:
```diff
@@ -1518,13 +1518,17 @@ def install(name=None,
+                    # in ver, otherwise its a dict.
+                    ver2 = ver['version'] if diff_attr else ver
+                    ver2 = norm_epoch(ver2, version_num)
+                    if salt.utils.compare_versions(ver1=version_num,
```
`salt.utils.versions.compare` changed to `salt.utils.compare_versions` @brejoc

brejoc committed:
Switching back to new usage of utils lib

Old usage slipped in and this reverts it back to the new usage.

brejoc commented:
Hi @terminalmage, would you please be so nice to take a look here, so that we can proceed. Thanks!

terminalmage commented on salt/modules/yumpkg.py:
```diff
@@ -1518,13 +1518,17 @@ def install(name=None,
+                    # if diff_attr was provided, version number is directly
+                    # in ver, otherwise its a dict.
+                    ver2 = ver['version'] if diff_attr else ver
+                    ver2 = norm_epoch(ver2, version_num)
```
This will not handle epochs properly. In fact, I dislike the entire way `diff_attr` was implemented, now that I think about it, since it destroys the information we get from pkg.list_pkgs. Please hold off on this, I have changes to make to how `diff_attr` works in the oxygen branch, which may require that this PR be updated or even unnecessary.

terminalmage commented:
@brejoc can you have a look at https://github.com/saltstack/salt/pull/45275? This should make it so that the version is handled properly when `diff_attr` is used. That should also make this PR unnecessary, though I wouldn't mind getting your tests in separately.

brejoc commented:
Closing in favour of https://github.com/saltstack/salt/pull/45275.

terminalmage commented:
@brejoc If you feel your tests would be beneficial, please feel free to open a new PR.
